cmd/internal/cli/pull.go

```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
 			sylog.Fatalf("While creating Docker credentials: %v", err)
 		}
 
-		libexec.PullOciImage(name, args[i], types.Options{
-			TmpDir:           tmpDir,
-			Force:            force,
-			NoHTTPS:          noHTTPS,
-			DockerAuthConfig: authConf,
-			NoCleanUp:        noCleanUp,
-		})
+		sysCtx := &ocitypes.SystemContext{
+			OCIInsecureSkipTLSVerify:    noHTTPS,
+			DockerInsecureSkipTLSVerify: noHTTPS,
+			DockerAuthConfig:            authConf,
+		}
+
+		sum, err := ociclient.ImageSHA(args[i], sysCtx)
+		if err != nil {
+			sylog.Fatalf("Failed to get SHA of %v: %v", args[i], err)
+		}
+
+		name := uri.GetName(args[i])
+		imgabs := cache.OciTempImage(sum, name)
+
+		exists, err := cache.OciTempExists(sum, name)
+		if err != nil {
+			sylog.Fatalf("Unable to check if %v exists: %v", imgabs, err)
+		}
+		if !exists {
+			sylog.Infof("Converting OCI blobs to SIF format")
+			b, err := build.NewBuild(
+				args[i],
+				build.Config{
+					Dest:   imgabs,
+					Format: "sif",
+					Opts: types.Options{
+						TmpDir:           tmpDir,
+						NoTest:           true,
+						NoHTTPS:          noHTTPS,
+						DockerAuthConfig: authConf,
+					},
+				},
+			)
+			if err != nil {
+				sylog.Fatalf("Unable to create new build: %v", err)
+			}
+
+			if err := b.Full(); err != nil {
+				sylog.Fatalf("Unable to build: %v", err)
+			}
+		} else {
+			sylog.Infof("Using cached image")
+		}
+
+		// Perms are 777 *prior* to umask
+		dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+		if err != nil {
+			sylog.Fatalf("Unable to open file: %s: %v\n", name, err)
+		}
+		defer dstFile.Close()
+
+		srcFile, err := os.OpenFile(imgabs, os.O_RDONLY, 0444)
```

Comment:
mask is not used when reading a file, this is the same as `os.Open(imgabs)`

Reply:
Can you explain a bit more?